kubeadm: simplify minimum Kubernetes version bumps

bart0sh opened the pull request:
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Replaced hardcoded "v0.12.0" strings with new
MinimalKubernetesVersion constant.

This should help with a regular release version bumps.

**Special notes for your reviewer**:

This is a first step towards fixing kubernetes/kubeadm#1260

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

bart0sh committed:
kubeadm: simplify minimum Kubernetes version bumps

Replaced hardcoded "v0.12.0" strings with MinimumControlPlaneVersion and
MinimumKubeletVersion global variables.

This should help with a regular release version bumps.

rosti commented on cmd/kubeadm/app/apis/kubeadm/v1alpha3/conversion_test.go:
```diff
@@ -112,7 +113,7 @@ func TestConvertToUseHyperKubeImage(t *testing.T) {
 			in: &ClusterConfiguration{
 				ImageRepository:          "k8s.gcr.io",
-				KubernetesVersion:        "v1.12.0",
+				KubernetesVersion:        constants.MinimumKubernetesVersion,
```
We don't need this one. The tested function does not depend on the current Kubernetes version.

bart0sh replied:
Thank you for the review. will revert this and all the following cases you've pointed out.

rosti commented on cmd/kubeadm/app/cmd/phases/util_test.go:
```diff
@@ -44,8 +45,8 @@ func TestSetKubernetesVersion(t *testing.T) {
-			input:  "v1.12.0",
-			output: "v1.12.0",
+			input:  constants.MinimumKubernetesVersion,
+			output: constants.MinimumKubernetesVersion,
```
Again, we don't need this to test `SetKubernetesVersion`. It can stay forever on v1.12.0.

rosti commented on cmd/kubeadm/app/componentconfigs/config_test.go:
```diff
@@ -40,7 +40,7 @@ kind: KubeletConfiguration
 
 func TestGetFromConfigMap(t *testing.T) {
-	k8sVersion := version.MustParseGeneric("v1.12.0")
+	k8sVersion := version.MustParseGeneric(kubeadmconstants.MinimumKubernetesVersion)
```
I don't think, that this is necessary either.

rosti commented on cmd/kubeadm/app/util/version_test.go:
```diff
@@ -429,12 +431,12 @@ func TestValidateStableVersion(t *testing.T) {
 		{
 			name:          "invalid: error parsing the client version",
-			remoteVersion: "v1.12.0",
+			remoteVersion: constants.MinimumKubernetesVersion,
```
`MinimumKubernetesVersion` is not needed to test the logic in `validateStableVersion`

rosti commented on cmd/kubeadm/app/constants/constants.go:
```diff
@@ -334,6 +334,9 @@ const (
 	CoreDNSVersion = "1.2.6"
 
+	// MinimumKubernetesVersion specifies minimum Kubernetes version supported by kubeadm
+	MinimumKubernetesVersion = "v1.12.0"
```
We can probably trim this to `v1.12`, so we can cover more cases with this PR (like `v1.12.5`).

bart0sh replied:
I'd prefer to do it in a follow-up PR as this one is already big and could cause rebase conflicts if not merged soon. wdyh?

neolit123 replied:
better to have it in the same PR, IMO.

bart0sh replied:
ok, will do

neolit123 replied:
but i wonder what other cases we need to do?

bart0sh replied:
@rosti, @neolit123 Would something like this be acceptable:
```
// MinK8sVerMajorMinor specifies <major>.<minor> part of the minimum Kubernetes version supported by Kubeadm

MinK8sMajorMinor = "v1.12"

// MinimumKubernetesVersion specifies minimum Kubernetes version supported by kubeadm
MinimumKubernetesVersion = MinK8sVerMajorMinor + ".0"
```

Then we can use MinK8sVerMajorMinor constant in the rest of v1.12* strings.

neolit123 replied:
on each version bump we need to send a PR that does something like this:
https://github.com/kubernetes/kubernetes/pull/68649/files

the idea is to reduce this to a change in a single file - constants.go.

bart0sh replied:
I think I understood the idea. I don't understand why it has to be implemented as one PR.

neolit123 replied:
my vote would be for:
```
MinimumKubernetesVersionWithoutPatch = ...
MinimumKubernetesVersion = ...
```

in terms of the example change in:
https://github.com/kubernetes/kubernetes/pull/68649/files

we might as well add more constants for the sake of making it more easy.
e.g. Minor + 1

that's my 2c.

rosti replied:
I was thinking about having all changes, that involve `1.12.x` in a single PR (currently we replaced only `1.12.0` and left some `1.12.x` stuff, where x != 0, for another PR).

I am OK with moving `1.13` and `1.14` stuff in other PR(s). That way it will be more reviewable.

bart0sh replied:
@rosti 

> currently we replaced only 1.12.0 and left some 1.12.x stuff

Nope, almost everything is covered with MinimumKubernetesVersion and MinimumKubernetesVersionWithoutPatch. You probably missed my updates.

Below is what's left currently. Can you point me out what of this should be included into this PR?
```
$ grep -r v1.12 . | grep -v ':\s*\/\/'
Binary file ./kubeadm matches
./app/constants/constants.go:	MinimumKubernetesVersionWithoutPatch = "v1.12"
Binary file ./app/cmd/phases/phases.test matches
./app/cmd/phases/util_test.go:			input:  "v1.12.0",
./app/cmd/phases/util_test.go:			output: "v1.12.0",
./app/cmd/config_test.go:				":v1.12.1",
./app/cmd/config_test.go:				kubernetesVersion: v1.12.1
./app/componentconfigs/config_test.go:	k8sVersion := version.MustParseGeneric("v1.12.0")
./app/phases/uploadconfig/uploadconfig_test.go:					KubernetesVersion: "v1.12.10",
Binary file ./app/util/upgrade/upgrade.test matches
./app/util/config/testdata/conversion/master/v1alpha3.yaml:kubernetesVersion: v1.12.2
./app/util/config/testdata/conversion/master/v1alpha3.yaml:unifiedControlPlaneImage: "k8s.gcr.io/hyperkube:v1.12.2"
./app/util/config/testdata/conversion/master/v1beta1.yaml:kubernetesVersion: v1.12.2
./app/util/config/testdata/conversion/master/internal.yaml:KubernetesVersion: v1.12.2
./app/util/config/common.go:		"kubeadm.k8s.io/v1alpha2": "v1.12",
./app/util/version_test.go:			remoteVersion: "v1.12.0",
./app/util/version_test.go:			remoteVersion: "v1.12.1",
./app/util/version_test.go:			output:        "v1.12.1",
./app/util/version_test.go:			clientVersion: "v1.12.0",
./app/util/version_test.go:			remoteVersion: "v1.12.0",
./app/util/etcd/etcd.go:				klog.V(1).Infoln("etcd manifest created by kubeadm v1.12")
./app/apis/kubeadm/v1alpha3/conversion_test.go:				KubernetesVersion:        "v1.12.2",
./app/apis/kubeadm/v1alpha3/conversion_test.go:				UnifiedControlPlaneImage: "k8s.gcr.io/hyperkube:v1.12.2",
./app/apis/kubeadm/v1alpha3/conversion_test.go:				KubernetesVersion:        "v1.12.0",
./app/apis/kubeadm/v1alpha3/conversion_test.go:				UnifiedControlPlaneImage: "k8s.gcr.io/hyperkube:v1.12.2",
./app/apis/kubeadm/v1alpha3/conversion_test.go:				KubernetesVersion:        "v1.12.2",
./app/apis/kubeadm/v1alpha3/conversion_test.go:				UnifiedControlPlaneImage: "k8s.gcr.io/hyperkube:v1.12.2",
./app/apis/kubeadm/v1alpha3/conversion_test.go:				KubernetesVersion:        "v1.12.2",
./app/apis/kubeadm/v1alpha3/conversion_test.go:				UnifiedControlPlaneImage: "k8s.gcr.io/otherimage:v1.12.2",
```

rosti replied:
Sorry @bart0sh , I missed the updates. Will take a look.

rosti reviewed (commented):
Thanks for taking this up @bart0sh !
We can cover a few more cases with this PR (all `v1.12.x` cases).
A few of the spots don't depend on the current version. It's OK to keep them wired to a string as that may make code more readable.

rosti commented:
/priority important-longterm
/assign @neolit123

neolit123 reviewed (approved):
i think this is good.
thanks @bart0sh

neolit123 commented:
/approve
/hold
/assign @fabriziopandini 
for another set of eyes.

rosti commented:
/test pull-kubernetes-godeps

bart0sh commented:
/test pull-kubernetes-godeps

rosti reviewed (approved):
Thanks @bart0sh ! Just a single one missing:
https://github.com/kubernetes/kubernetes/blob/1c04caa04325e1f64d9a15714ad61acdd2a81013/cmd/kubeadm/test/cmd/init_test.go#L105

Then we are ready to LGTM the 1.12.x portion.

bart0sh commented:
@rosti Thank you for the review. Updated. lgtm, please as I'm concerned about possible conflicts.

rosti commented:
Thanks again @bart0sh !
/lgtm
/test pull-kubernetes-e2e-kops-aws

bart0sh commented:
/test pull-kubernetes-e2e-kops-aws

fabriziopandini reviewed (commented):
@bart0sh 
Thanks for this PR! simplify Kubernetes version bumps will be of great help!

One background question for this PR. Have you considered to implement a more declarative approach for replacing all the "version string concatenation magic" currently existing in kubeadm tests? e.g.
 
```
func (v Version) WithMinor(patch uint) *Version {
}

func (v Version) WithPatch(patch uint) *Version {
}

func (v Version) WithPreRelease(prerelease string) *Version {
}
```

Or something similar for getting variations of the `MinimumKubernetesVersion `?

bart0sh commented:
@fabriziopandini Thank you for the idea. So far I don't see that ```constants.MinimumKubernetesVersion + "-alpha.1``` is much worse than  ```constants.MinimumKubernetesVersion.WithPatch(0-alpha.1)```, but can play with it to see if it makes sense. 

BTW, does anybody know about existing Kubernetes APIs in this area?

bart0sh committed:
version: add 3 methods

Added WithMajor, WithMinor, WithPatch and WithPreRelease methods
to the Version API.

These methods return copy of the Version object with one changed
property(Major, Minor, Patch or preRelease).

rosti commented on cmd/kubeadm/app/cmd/alpha/kubelet.go:
```diff
@@ -38,13 +40,13 @@ var (
-		kubeadm alpha phase kubelet config download --kubelet-version v1.12.0
-		`)
+		kubeadm alpha phase kubelet config download --kubelet-version %s
+		`, constants.MinimumKubeletVersion.String()))
```
No need for the `.String()` portion here, `Sprintf` will do it for you.

rosti commented on cmd/kubeadm/app/cmd/alpha/kubelet.go:
```diff
@@ -55,13 +57,13 @@ var (
 		WARNING: This feature is still experimental, and disabled by default. Enable only if you know what you are doing, as it
 		may have surprising side-effects at this stage.
-		`)
+		`, constants.MinimumKubeletVersion.String()))
```
And here.

rosti commented on cmd/kubeadm/app/cmd/config_test.go:
```diff
@@ -83,11 +85,11 @@ func TestImagesListRunWithCustomConfigPath(t *testing.T) {
-				kubernetesVersion: v1.12.0
-			`)),
+				kubernetesVersion: %s
+			`, constants.MinimumControlPlaneVersion.String()))),
```
And here.

rosti commented on cmd/kubeadm/test/util.go:
```diff
@@ -66,8 +67,8 @@ func SetupInitConfigurationFile(t *testing.T, tmpdir string, cfg *kubeadmapi.Ini
-		kubernetesVersion: v1.12.0
-		`)))
+		kubernetesVersion: %s
+		`, kubeadmconstants.MinimumControlPlaneVersion.String()))))
```
And here.

bart0sh commented:
@neolit123, @fabriziopandini, @rosti I've replaced string concatenations with using With* methods as @fabriziopandini suggested. Please, review again. Thanks.

neolit123 commented:
@bart0sh 
the `staging/src/k8s.io/apimachinery/pkg/util/version/version.go` helper changes LGTM, but FYI this can drag this PR a lot in time as it enters the scope of another SIG.

rosti reviewed (approved):
Thanks for the update @bart0sh !
Actually I like this approach better, but as @neolit123 we need to sell the idea to SIG API Machinery and that could take time or even not happen at all.
A few unnecessary `String()` calls can be omitted, but I can live with them.
I give you LGTM on the kubeadm side of things.
/lgtm

bart0sh commented:
@rosti, @neolit123 Thank you for the review!

As for the change of Version API, I tried to avoid it in the previous version of this patch. Then @fabriziopandini suggested this approach and I reworked the patch. I still have previous version, so please suggest which way we want to go.

In my opinion we should go faster way(previous patch) and submit Version API change as a separate PR. When/if it's accepted I'll be happy to submit new PR for kubeadm code that uses changed API.

Thoughts?

fabriziopandini reviewed (approved):
@bart0sh thanks for this news version!
/approve
/lgtm

Let's wait for sig-api-machinery opinion

bart0sh commented:
@fabriziopandini @neolit123 Is anything required from me? Should I contact someone or should we assign someone from sig-api-machinery to this PR?

neolit123 commented:
@bart0sh the likelyhood of this making it in before NY is minimal due to the api-machinery change.

/assign @caesarxuchao

bart0sh commented:
@neolit123 so, should we wait or use previous version of this PR and make separate PR for api change as I proposed earlier?

neolit123 commented:
@bart0sh 

as per @fabriziopandini 's comment:
>Let's wait for sig-api-machinery opinion

neolit123 commented on cmd/kubeadm/app/cmd/alpha/kubelet.go:
```diff
@@ -38,13 +40,13 @@ var (
-		kubeadm alpha phase kubelet config download --kubelet-version v1.12.0
-		`)
+		kubeadm alpha phase kubelet config download --kubelet-version %s
+		`, constants.MinimumKubeletVersion))
```
@fabriziopandini @rosti @bart0sh @rojkov

instead of the usage of this constant here i think we should just have a `CurrentKubernetesMinorVersion` (or a similar naming) that is `v1.13.0`.
we can then use this version in such CLI examples and also have it as the current version for unit tests.
WDYT?

liggitt replied:
would that `Current...` constant require manual updating on every release? that does not seem like a good idea

neolit123 replied:
we already bump these on each release:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/constants/constants.go#L380-L383

in an ideal world:
- an app runtime should fail if the build system did not add versioning to the binary.
- skew versions and current versions should be obtained in variables based of the client version.
- unit tests should not be bound to internet or client versioning functionality.
- a default version should not be fetched from the internet by default.

fabriziopandini replied:
Considering that this PR IMO is already achieving the goals to simplify version bumps, which was the original scope of the underlying issue, I'm not sure it is a good idea to conflate in this PR other goals like fixing version discovery for unit test or shifting to a dynamic approach for handling version skew policy. Might be follow up PRs are more appropriate in this case

neolit123 replied:
then again, using `constants.MinimumKubeletVersion` as an example in the CLI might introduce problems. this is a version that is currently 1.12 instead of 1.13.

rosti replied:
The idea of `CurrentKubernetesVersion` is good one, but outside of the scope of this PR. It can be done in some of the followup PRs though. For a start we can bump this version manually at the end of the release cycle, but ideally this should be done automatically by the build system.

rosti commented:
I think, that we can safely cancel the hold now and let this merge.
@neolit123 @fabriziopandini WDYT?

bart0sh commented:
in my opinion it's ok to merge it. If there are any issues with it I'll fix it. It's reviewed and approved by many people already.

neolit123 commented:
/hold cancel

bart0sh commented:
/retest
